### ISSUE TITLE
Classic gotcha of   mismatched input 'xyz' expecting 'xyz'

### ISSUE THREAD

**Issue by SimonSntPeter** [antlr/antlr4#1072]
Did the usual newbie mistake of defining identifiers too high up in the file so got errors of this form. Definitely not alone in this (q.v. <http://stackoverflow.com/questions/29777778/antlr-4-5-mismatched-input-x-expecting-x>), and it's very confusing for noobs.
I'd suggest in general that the token type always be mentioned, something like

    mismatched input IDENT 'begin' expecting PROGRAM

and in this special case where the given strings are the same (eg. mismatched input 'with' expecting 'with') that some extra info be reported, perhaps state directly, or give reference to this in the ANTLR4 faq, explaining that lexing is done in physical order of rules in the file and you've probably got it wrong.

Quoting from Sam Harwell "The Vocabulary interface [...] provides the necessary information to dramatically improve reporting for [these examples]. [Suggest creation of] a new error handler (possibly as an example in the documentation) which uses information from the Vocabulary along with the exception details to provide more actionable feedback for developers."

Question - "possibly as an example in the documentation" - does this mean that it would not be the default error handler? Because that would defeat the point, as newcomers (who have most to gain from it) would not know how to invoke it, or even of its presence.

**Comment by KvanTTT** [comment]
You are right, warning should be generated.
More simple example:
```ANTLR
Test1: 'A';
// ...
// several token declarations
// ...
Test2: 'A';
```

**Comment by sharwell** [comment]
@KvanTTT That case is possible, but it's more common for something like this to occur.

```antlr
ID : [A-Z]+;
Test2 : 'A';
```

No `Test2` token will ever be created, yet it's not possible to statically determine whether or not two arbitrary rules overlap. More importantly, when you reverse the order of these rules you still have a conflict for the input `A`. The only difference is which token type is assigned in the presence of an ambiguity.

```antlr
Test2 : 'A';

// ID will never be assigned for the input A, but unlike the previous example this
// is now desirable.
ID : [A-Z]+;
```

**Comment by KvanTTT** [comment]
@sharwell, as I understand this is problem about DFA intersection? If resulted DFA is not null than tokens may match equal lexems and warning should be throwed.

**Comment by SimonSntPeter** [comment]
@KvanTTT - @sharwell's point is that such intersection is necessary and unavoidable, and is disambiguated by precedence (physical position in the file for ANTLR). See his example immediately above your post. 
As identifiers is typically [a-zA-Z0-9_]+, but that matches 'SELECT' and 'JOIN' and 'WHERE' and ... everything really. Would you want a warning in these cases?
The only solution would be to have some regex extension which says "this but not these", perhaps something like 
```[a-zA-Z0-9_]+ \ ( 'SELECT' | 'JOIN' | 'WHERE' ) ```
where '\' indicates set exclusion. Immediate thought is the you could extend that to lexer rules for the right hand side: 
```[a-zA-Z0-9_]+ \ RESERVED_WORD``` 
My second immediate thought is that while it rather appeals to my neat and tidy mind, I don't see it happening.

**Comment by KvanTTT** [comment]
In this case the warning (or error) should be generated (because of 'SELECT' is never reachable).
```ANTLR
ID: [A-Z]+;                // or SELECT_0: 'SELECT';
SELECT: 'SELECT';
```
But there is a normal case because of ID is desirable:
```ANTLR
SELECT: 'SELECT';
ID: [A-Z]+;
```

**Comment by KvanTTT** [comment]
I also encountered with problem "mismatched input 'xyz' expecting 'xyz'" when lexer file has been changed without parse.

My suggestion: If token text (lexem) matches one of the expected literal name, than add token type to message. Something like this:
```mismatched input 'xyz' (type: 123) expecting 'xyz' (type: 456)```

**Comment by parrt** [comment]
I think something akin to ambiguity detection in parser should be implemented in the lexer. These can therefore be reported dynamically.

**Comment by KvanTTT** [comment]
Some cases can be detected statically. On my experience, it's pretty common pattern of errors. I suggest to implement static checks for exactly duplicate literals:
```
Token1: 'A';
// ...
Token2: 'A';
```
and for conflicts of ID and other tokens:
```
ID: [a-z]+;
// ...
TOKEN1: 'A';
```

**Comment by KvanTTT** [comment]
>A cursory look indicates it doesn't find keywords shadowed by ID's however.

Yes, my solution from #2066 finds warnings for only simple cases with duplicated tokens. Deeper analysis requires at least DFA's difference algorithm implementation.

The general satisfactory solution will check a difference between token's DFA and throw a warning if it's not an empty, i.e. `A \ B != Ø` where `A` is a token declaration written before the `B` token declaration.

Let check this term in practice.

This sample will throw a warning because of difference between `ID` rule and `A` rule is `[B-Z][A-Z]*`, i.e. not an empty. As [I already said](https://github.com/antlr/antlr4/issues/1072#issuecomment-302770116) it is a very common pattern of errors.

```antlr
ID : [A-Z]+;
A : 'A';
```

This sample won't throw a warning because of difference is an empty language.

```antlr
A : 'A';
ID : [A-Z]+;
```

The following sample also will throw a warning:

```antlr
A : 'A';
B : 'A';
```

Despite the fact that ANTLR lexer covers not only regular languages but [context-free](https://github.com/antlr/antlr4/blob/master/doc/lexer-rules.md#recursive-lexer-rules) too I suggest checking only not recursive tokens. Those exclude tokens like this:

```
ACTION : '{' ( ACTION | ~[{}] )* '}' ;
```

Such tokens are rare and do not significantly affect warnings checking.

**Comment by manjula91** [comment]
But what if I want something like this to work

```
A : 'A';
B : 'A';

```

**Comment by exander77** [comment]
I have a question related to this issue, it is possible to tell Antlr to create multiple lexers for different parts of the grammar?

**Comment by KvanTTT** [comment]
> I have a question related to this issue, it is possible to tell Antlr to create multiple lexers for different parts of the grammar?

You can try to use [lexical modes](https://github.com/antlr/antlr4/blob/master/doc/lexer-rules.md#lexical-modes).
